orc8r/cloud/go/services/ctraced/obsidian/handlers/handlers.go

```diff
@@ -0,0 +1,246 @@
+		if err != nil {
+			return obsidian.HttpError(errors.Wrap(err, "failed to create call trace"), http.StatusInternalServerError)
+		}
+		return c.JSON(http.StatusCreated, string(cfg.TraceID))
+	}
+}
+
+func GetCallTrace(c echo.Context) error {
+	callTrace, err := getCallTrace(c)
+	if err != nil {
+		return err
+	}
+	return c.JSON(http.StatusOK, callTrace)
+}
+
+func GetUpdateCallTraceHandlerFunc(client GWCtracedClient, storage storage.CtracedStorage) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		networkID, callTraceID, nerr := getNetworkIDAndCallTraceID(c)
+		if nerr != nil {
+			return nerr
+		}
+		mutableCallTrace := &models.MutableCallTrace{}
+		if err := c.Bind(mutableCallTrace); err != nil {
+			return obsidian.HttpError(err, http.StatusBadRequest)
+		}
+		if err := mutableCallTrace.ValidateModel(); err != nil {
+			return obsidian.HttpError(err, http.StatusBadRequest)
+		}
+
+		callTrace, err := getCallTrace(c)
+		if err != nil {
+			return err
+		}
+
+		req := &protos.EndTraceRequest{}
+		resp, err := client.EndCallTrace(networkID, callTrace.Config.GatewayID, req)
+		if err != nil {
+			return err
+		}
+		if resp.Success == false {
+			return obsidian.HttpError(errors.New("Failed to end call trace"), http.StatusInternalServerError)
+		}
+
+		err = storage.StoreCallTrace(networkID, callTraceID, resp.TraceContent)
+		if err != nil {
+			return obsidian.HttpError(errors.Wrap(err, "failed to save call trace data"), http.StatusInternalServerError)
+		}
+
+		_, err = configurator.UpdateEntity(networkID, mutableCallTrace.ToEntityUpdateCriteria(callTraceID, *callTrace), serdes.Entity)
+		if err != nil {
+			return obsidian.HttpError(err, http.StatusInternalServerError)
+		}
+		return c.NoContent(http.StatusNoContent)
+	}
+}
+
+func GetDeleteCallTraceHandlerFunc(client GWCtracedClient, storage storage.CtracedStorage) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		networkID, callTraceID, nerr := getNetworkIDAndCallTraceID(c)
+		if nerr != nil {
+			return nerr
+		}
+
+		err := storage.DeleteCallTrace(networkID, callTraceID)
+		if err != nil {
+			return obsidian.HttpError(errors.Wrap(err, "failed to delete call trace data"), http.StatusInternalServerError)
+		}
+
+		err = configurator.DeleteEntity(networkID, orc8r.CallTraceEntityType, callTraceID)
+		if err != nil {
+			return obsidian.HttpError(err, http.StatusInternalServerError)
+		}
+		return c.NoContent(http.StatusNoContent)
+	}
+}
+
+func GetDownloadCallTraceHandlerFunc(storage storage.CtracedStorage) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		networkID, callTraceID, nerr := getNetworkIDAndCallTraceID(c)
+		if nerr != nil {
+			return nerr
+		}
+
+		callTrace, err := storage.GetCallTrace(networkID, callTraceID)
+		if err != nil {
+			return obsidian.HttpError(errors.Wrap(err, "failed to retrieve call trace data"), http.StatusInternalServerError)
+		}
+
+		return c.Blob(http.StatusOK, "application/pcap", callTrace)
```

Comment:
is tshark file of pcap format or pcapng format ?

Reply:
Oh it's pcapng actually